File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        charMap = {
            '(' : -1,
            '{' : -2,
            '[' : -3,
            ')' : 1,
            '}' : 2,
            ']' : 3
        }
        
        stack = []
        
        for i in range(0, len(s)):
            c = charMap.get(s[i])
            
            if len(stack) > 0:
                top = stack.pop()

                if c == -1 * top:
                    continue
                elif c < 0:
                    stack.append(top)
                    stack.append(c)
                else:
                    return False
            else:
                if c > 0:
                    return False

                stack.append(c)
        
        return len(stack) == 0

File: 0020-valid-parentheses/0020-valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        unordered_map<char, int> charMap = {
            {'(', -1},
            {'{', -2},
            {'[', -3},
            {')', 1},
            {'}', 2},
            {']', 3}
        };
        
        stack<int> stk;
        
        for (char c : s) {
            
            int n = charMap[c];
            
            if (stk.size() > 0) {
                
                if (n == -1 * stk.top())
                    stk.pop();
                else if (n < 0)
                    stk.push(n);
                else
                    return false;
                
            }
            else {
                
                if (n > 0)
                    return false;
                
                stk.push(n);
                
            }
            
        }
        
        return stk.size() == 0;
    }
};